### PR TITLE
Added `rustfmt_toolchain` and refactored toolchain repository rules

### PR DESCRIPTION
**Pull request by UebelAndre**
This change introduces a `rustfmt_toolchain` and refactors repository rules to create this toolchain over adding `rustfmt` to the standard `rust_toolchain`. This new toolchain saves time and disk space since in many cases `rustfmt` is not required for compiling. 

Additionally, `rust_register_toolchains` has been refactored to deprecate `version` and `iso_date`. Instead, this information can be conveyed through `versions` and `rustfmt_version` by passing variables such as the following `["1.66.0", "nightly/2022-12-15"]`.

### CODE REVIEW

File: rust/private/repository_utils.bzl [hunk 1]
```diff
@@ -730,3 +748,44 @@ def _get_tool_extension(ctx):
...
+
+def select_rust_version(versions):
+    """Select the highest priorty version for a list of Rust versions
+
+    Priority order: `stable > nightly > beta`
+
+    Note that duplicate channels are unexpected in `versions`.
+
+    Args:
+        versions (list): A list of Rust versions. E.g. [`1.66.0, `nightly/2022-12-15`]
```
Comment (krasimirgg) [inline rust/private/repository_utils.bzl:760]: nit: missing backtick between the `0` and the `,`
```
E.g. [`1.66.0`
```

**Review by krasimirgg — APPROVED** [review]
Looks good! Thank you!

File: rust/private/repository_utils.bzl (hunk 1 above)
Comment (UebelAndre) [reply rust/private/repository_utils.bzl:760]: Thanks! Fixed!

**Comment by UebelAndre** [comment]
@illicitonion any concerns?

**Review by illicitonion — APPROVED** [review]

**Pull request merged**